mlir/include/mlir/Dialect/MIOpen/AffineMapHelper.h

```diff
@@ -0,0 +1,75 @@
+//===- AffineMapHelper.h - Utility routines for AffineMap ---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provide utility routines to check AffineMap instances.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_MIOPEN_AFFINEMAP_HELPER_H
+#define MLIR_DIALECT_MIOPEN_AFFINEMAP_HELPER_H
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Attributes.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace miopen {
+
+//===----------------------------------------------------------------------===//
+// Check if an AffineMap has division or remainder inside.
+//===----------------------------------------------------------------------===//
+inline bool hasDivisionOrRemainder(AffineMap map) {
+  bool ret = false;
+  if (!map)
+    return false;
+  map.walkExprs([&ret](AffineExpr expr) {
+    if (expr.getKind() == AffineExprKind::Mod ||
+        expr.getKind() == AffineExprKind::FloorDiv ||
+        expr.getKind() == AffineExprKind::CeilDiv)
+      ret = true;
+  });
+
+  // XXX. hack. always return false for now for performance reason.
+  // May need more sophisticated checks to determine if we would truly go OOB.
+  // return ret;
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// Check if an AffineMap has padding, which is represented as a minus expression
+// with a constant operand.
+//===----------------------------------------------------------------------===//
+inline bool hasPadding(AffineMap map) {
```

Comment:
I think instead of implementing ```hasPadding()``` for ```AffineMap map```, it's better to implement ```needOutOfRangeCheck()``` for each ```AffineExpr expr```. If any ```AffineExpr expr``` need out-of-range check, then the whole ```AffineMap map``` need IsValid check

Reply:
@asroy It will be in the subsequent PR. The commit is here:

https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/commit/e068a6e7a5544928a05799776b56e1a0f25e0382

Reply:
e068a6e looks good.